i_1077 Handle each EF client in its own thread

johnbrvc opened the pull request:
### Description of what the PR does
This will prevent one bad client from clogging up the `EventFeedStreamer`.
Also write new event to log FIRST so we don't get duplicate events in the event that PC2 is stopped while sending to clients.
Retry submission fetching.

### Issue which the PR addresses
Fixes #1077 
### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

1. This bug fix is very hard to test.
2. You need a full contest.
3. You need playback data for the contest (via feeder1)
4. You need to start up a 2023-06 API server (via feeder2)
5. You need to start several shadows to the primary (3 shadows should do)
6. You need to start 2 CDS's to the primary.
7. You then have to "jam -up" one or more of the shadows/CDS's by blocking their firewall or unplugging their network cable.  This will cause the connection to hang on the PC2 Event Feeder side.
8. PC2 should recover from these type of connected client failures.

johnbrvc committed:
i_1077 Handle each EF client in its own thread

This will prevent one bad client from clogging up the EventFeedStreamer.
Also write new event to log FIRST so we dont get duplicate events in the event that PC2 is stopped while sending to clients.
Retry submission fetching.

johnbrvc committed:
i_1077 Fix JUnit - output has to complete for EF JSON

The JUnit exposed a potential problem with prematurely removing an output stream from an EventFeedStreamer. With the new threaded implementation, it's possible that data was still "in the pipe" waiting to be flushed.  We have to wait for that.
Also make the Junit failure message a bit more descriptive.

troy2914 commented:
This looks reasonable to me, though i only did a code review.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java:
```diff
@@ -128,24 +130,83 @@ public class EventFeedStreamer extends JSON202306Utilities implements Runnable,
+        }
+
+        public void setTerminateClient() {
+            // if terminateClient is false, this means he caller is asking us
+            // to stop because they're done.  (eg. removeClient() being called)
+            // If there's data pending, we want to wait a big for it to drain.
```
Multiple minor typos in the above lines:
- `he` was probably intended to be `the`
- `big` was probably intended to be `bit`.

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java:
```diff
@@ -128,24 +130,83 @@ public class EventFeedStreamer extends JSON202306Utilities implements Runnable,
+
+        public void setTerminateClient() {
+            // if terminateClient is false, this means he caller is asking us
+            // to stop because they're done.  (eg. removeClient() being called)
```
It's not clear what the reference to `removeClient()` in the above comment is talking about.  As far as I can find, there is no such method anywhere in the code.  (Was this intended to refer to `removeStream()`?)

johnbrvc replied:
Yes, removeStream() is what was meant.  Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java:
```diff
@@ -128,24 +130,83 @@ public class EventFeedStreamer extends JSON202306Utilities implements Runnable,
+            // if terminateClient is false, this means he caller is asking us
+            // to stop because they're done.  (eg. removeClient() being called)
+            // If there's data pending, we want to wait a big for it to drain.
+            if(terminateClient == false && outputQueue.isEmpty() == false) {
```
I don't understand the purpose of the `terminateClient == false` portion of the above `if`.  If someone calls this method (`setTerminateClient()`), and the outputQueue is not empty, don't we want to execute the `Thread.sleep()` regardless of whether `terminateClient` has already been set true?

johnbrvc replied:
No.  If `terminateClient `is already true, it means an error happened (we're dead already).  See the `catch`'s in the` run() `method.  In case of an error, there is no point in waiting for the connection to drain.  I fixed the comment to indicate this.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java:
```diff
@@ -864,31 +929,32 @@ public synchronized void sendJSON(String string) {
 
         string = replaceEventId (string, newId);
 
+        // Add notification to log before sending to clients
```
It might be nice (i.e., to help future developers) if this comment include an explanation of WHY we are now putting the event in the log BEFORE sending it instead of after sending it.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java:
```diff
@@ -864,31 +929,32 @@ public synchronized void sendJSON(String string) {
+            // Check if there was a problem with this client, if so, we are no longer
+            // interested in providing it with data
+            if (streamAndFilter.wantTerminateClient()) {
+                removeStream(streamAndFilter.getStream());
```
The method `removeStream()` called in the above line is modifying the same map which is being used to control the surrounding `for` loop.  I think this could potentially lead to throwing a `ConcurrentModificationException`.     Unless you're convinced for some reason that no such `ConcurrentModificationException` can happen, I'd suggest handling removal outside the main `for` loop.  (I note that method `removeStream()` is also called directly from the `EventFeedStreamer.run()` method, so that appears to me to be one place where concurrent modification could happen...)

(The standard solution to such an issue is to keep a list within the `for` loop of those items (`streams`, in this case) which need to be removed, and then run a separate `for` loop (after this one finishes) to remove the desired streams. )

johnbrvc replied:
The thing being operated on and modified is a `ConcurrentHashMap<>`.  AFAICT it's safe to do things on this from different threads, and even in an iterator.   (https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) - that seems to indicate it will never throw a `ConcurrentModificationException`.  While the iterator may not be "up-to-date", that's safe even if it gets something that was removed from the map before we iterated to it, since the reference to the object is still valid.  While I do not think it is necessary, since we're using entrySet(), it may not _hurt_ to put a test for `null `in the loop.  Next push.

clevengr replied:
Ok, that _probably_ covers it.

clevengr reviewed (changes requested):
Reviewed `src/edu/csus/ecs/pc2/clics/API202306/EventFeedStreamer.java`.  In addition to several minor comments, I noted two places where I have questions about whether the code is correct.  I'll be happy accept a fix, or an explanation of why my comments don't apply.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/SubmissionService.java:
```diff
@@ -309,10 +310,11 @@ public synchronized Response getSubmissionFiles(@Context SecurityContext sc, @Pa
                 }
 
-                //if runFiles is still null we failed to get the runfiles from the local model
-                if (runFiles == null) {
+                // We try to fetch the runfiles a few times
+                for(int nTry = 1; runFiles == null && nTry <= MAX_RUNFILE_RETRIES; nTry++) {
+                    // we failed to get the runfiles from the local model
```
I do not understand the deletion of the line `if (runFiles == null)`.  The code just above that line attempts to pull the runfiles from the local model.  The previous version of the code only tries to fetch them from the server if it failed to get them from the local model (which is indicated by `runFiles` still being null).  Now, it looks like the code attempts to fetch the runFiles from the server (up to three times) _whether or not it successfully got them from the local model_.  This doesn't seem logical; it should only invoke the server IF it failed to get them locally.  Am I missing something?

johnbrvc replied:
Check out the `for()` loop just below.   It first checks if `runFiles == null`.  if `runFiles `is not null, the loop is never executed, and, as soon as it finds the `runFiles`, the loop will exit, else the sentinel will kick in.

clevengr replied:
Aha -- I missed the compound condition in the `for` loop.

clevengr reviewed (changes requested):
Reviewed `src/edu/csus/ecs/pc2/clics/API202306/SubmissionService.java`.  Marked one place where I do not understand the change to the code.

clevengr reviewed (changes requested):
Completed a code review.  Marked several places where I've requested either a fix, or an explanation of why the code changes don't appear correct to me.    Moving on to runtime testing.

johnbrvc committed:
i_1077 PR Changes requests

Fix spelling errors in comments.
Enhance comments to be more detailed.
Add null checks in iterator loops out of an abundance of caution.

SamanwaySadhu reviewed (approved):
Did only code review. LGTM!
